### PR TITLE
feat: Convert CumeDist to UDWF

### LINKED ISSUE #12695: Convert `BuiltInWindowFunction::CumeDist` to a user defined window function
### Is your feature request related to a problem or challenge?

Part of https://github.com/apache/datafusion/issues/8709

There is now no difference between "built in" / "prepackaged" scalar and aggregate functions in DataFusion, however there are still some ["built in" window functions ](https://github.com/apache/datafusion/blob/9b4f90ad1eefabdc0d5bbbfd99e58765b041bb77/datafusion/expr/src/built_in_window_function.rs#L42-L69)-- see the current source for [BuiltInWindowFunction](https://github.com/apache/datafusion/blob/main/datafusion/expr/src/built_in_window_function.rs) for the up to date list of what remains

The problem with having two different kinds of window functions is
1. There are some features that may not be available to User Defined Window Functions that rely on built in
2. Users can not easily choose which window functions to include or override the behavior if they need something different

### Describe the solution you'd like

I would like to remove the "built in" version of this function and convert it to a user defined function

### Describe alternatives you've considered

At a high level the process is:

1. Add a new [WindowUDFImpl](https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.WindowUDFImpl.html) in the [functions-window](https://github.com/apache/datafusion/tree/main/datafusion/functions-window) crate
2. Port the code from the relevant existing implementation of the the built in functions in [datafusion/physical-expr/src/window](https://github.com/apache/datafusion/tree/main/datafusion/physical-expr/src/window)
3. Remove the BuiltInWindowFunction variant and then get everything to compile (the compiler will show you where the existing implementations are)

### Additional context

Here are some good examples:
- Port of RowNumber in [Convert built-in row_number to user-defined window function #12030](https://github.com/apache/datafusion/pull/12030)

### PR DESCRIPTION
**Pull request by jonathanc-n**
## Which issue does this PR close?

Closes #12695 .

## What changes are included in this PR?
Converted CumeDist to user defined window function.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/window.slt [hunk 1]
```diff
@@ -1135,8 +1135,8 @@ SELECT
 query IRR
 SELECT
   c8,
-  CUME_DIST() OVER(ORDER BY c9) as cd1,
-  CUME_DIST() OVER(ORDER BY c9 ROWS BETWEEN 10 PRECEDING and 1 FOLLOWING) as cd2
+  cume_dist() OVER(ORDER BY c9) as cd1,
+  cume_dist() OVER(ORDER BY c9 ROWS BETWEEN 10 PRECEDING and 1 FOLLOWING) as cd2
```
Comment (findepi) [inline datafusion/sqllogictest/test_files/window.slt:1139]: necessary?

**Review by findepi — APPROVED** [review]

**Review by timsaucer — APPROVED** [review]
Thank you!

**Comment by jcsherin** [comment]
@jonathanc-n A roundtrip test for `cume_dist()` expression API can be added here:

https://github.com/apache/datafusion/blob/c22abb4ac3f1af8bbdf176ef0198988fc7b0982c/datafusion/proto/tests/cases/roundtrip_logical_plan.rs#L943-L952

File: datafusion/sqllogictest/test_files/window.slt (hunk 1 above)
Comment (jcsherin) [reply datafusion/sqllogictest/test_files/window.slt:1139]: It is consistent with conversion of other user-defined window functions.

File: docs/source/user-guide/sql/window_functions_new.md [hunk 2]
```diff
@@ -157,11 +157,20 @@ All [aggregate functions](aggregate_functions.md) can be used as window function
 
 ## Ranking Functions
 
+- [cume_dist](#cume_dist)
 - [dense_rank](#dense_rank)
 - [percent_rank](#percent_rank)
 - [rank](#rank)
 - [row_number](#row_number)
 
+### `cume_dist`
```
Comment (alamb) [inline docs/source/user-guide/sql/window_functions_new.md:166]: 🎉

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Love it -- thank you @jonathanc-n  and @findepi @timsaucer and @jcsherin  for the reviews

**Comment by alamb** [comment]
🚀

**Pull request merged**